### ISSUE TITLE
[iCloud] Bridge-Thing generates Communication Error

### ISSUE THREAD

**Issue by RalphSester** [openhab/openhab-addons#13282]
The Brigde-Thing (iCloud-Account) does not go online.

It gets the state: COMMUNICATION_ERROR with the following details information:

`java.util.concurrent.ExecutionException: org.eclipse.jetty.client.HttpResponseException: HTTP protocol violation: Authentication challenge without WWW-Authenticate header
`

The binding is not usable completly at this time.

## Current Behavior
The Bridge-Thing does not come online.
Restart of binding or OH doesn't help.

## Possible Solution
It seems, that apple has changed something in their api.

## Steps to Reproduce (for Bugs)
1. Just start the binding.
2. See the error

## Environment

* Version used: OpenHAB 3.3 / 3.4
* Operating System: Windows Server 2016

**Comment by BigGeorgeTx** [comment]
This is related to this https://github.com/openhab/openhab-addons/issues/13231 previously reported issue, but it appears to have gotten worse.  At least for me, A disable/enable cycle, restart of the binding, or restart of openHAB no longer fixes the issue.  From the comments in the openHAB community, https://community.openhab.org/t/icloud-binding-communication-error/122131/336?u=biggeorgetx, this is impacting many users.

**Comment by JensHoRi** [comment]
Maybe this information helps:
https://github.com/jetty-project/jetty-reactive-httpclient/issues/33

If I read it right (I' no developer), looks like apple changed something in authentication and is not compliant to specifications anymore.
Under that link, they had same issue and discussing a workaround. Looks like doing the authentication handling in code directly could be a temporary good solution. But I'm no developer, maybe I'm wrong. Just try to help

**Comment by BigGeorgeTx** [comment]
This issue returned today for many users. https://community.openhab.org/t/icloud-binding-communication-error/122131/327?u=biggeorgetx

**Comment by RalphSester** [comment]
It seems, that the iCloud-Binding has no developer since a couple of months.
Nobody seems to care, that the binding is completly unuseable.....

if this problem can't be fixed soon,  it would be better to remove it from OH 3.4 !!

@digitaldan: Do you have some ideas?

**Comment by martinvw** [comment]
I'm not very active currently as a maintainer, and will most likely not be so in the near future because of a planned move, but fixing such issues in an open-source project is not the sole responsibility of the binding maintainers, everyone is allowed, and should make a PR to fix this issue if they are capable to do so.

However, please note that in this specific case we are NOT using an official API provided by the company but relying on some internal API that was never intended for us to use. So it could well be that it is both hard to fix and only a temporary fix.

I do have a short amount of time now to spend on it and I do think that the suggested resolutions are not perse correct because the error tells us that Apple send us a 401 without including a `WWW-Authenticate`-header, this is not correct according to the spec but IMHO only a symptom of an underlying problem. 

We do send authentication in a so-called basic auth header, I just validated in Postman what happens when I perform such a request and it is actively killed by some service in between:

I receive the following response:

```
HTTP/1.1 401 
Server: AppleHttpServer/3faf4ee9434b
Date: Sun, 09 Oct 2022 09:18:05 GMT
Content-Length: 50
Connection: keep-alive
Set-Cookie: xr_3n2093n1a=HvkfmhuDkyiPeeoYx2YFIDS2DDiRlBJA==; Path=/; Secure; HttpOnly
X-Responding-Instance: fmipservice:3:2299B29512:2844136de5f5
X-Robots-Tag: noindex, noarchive, nosnippet, nofollow
X-Responding-Server: prod-p46-fmipservice--remaining-86888856bf-pm29h_99
X-Responding-Partition: p46
Strict-Transport-Security: max-age=31536000; includeSubDomains;
x-apple-user-partition: 46
via: xrail:mr47p00ic-qujn01123002.me.com:8301:22R479:grp21,631194250daa17e24277dea86cf30319:b2fb3d73b0bb3e9c280a2255c0ac93ae:nlhfd1
X-Apple-Request-UUID: b96e34d2-6603-4c85-b2d8-860866d4733b
access-control-expose-headers: X-Apple-Request-UUID,Via
X-Apple-Edge-Response-Time: 131

{"desc":"default text from ResponseContentFilter"}%
```

So my guess is that they might be explicitly banning us from using this old endpoint and the binding might have to be rewritten.

**Comment by JensHoRi** [comment]
It’s not a binding issue, apple not in line with http specs at the moment

**Comment by RalphSester** [comment]
> It’s not a binding issue, apple not in line with http specs at the moment

Source ?

**Comment by JensHoRi** [comment]
Al lot of research 6 month ago when same error was shown. You can look in the community too

**Comment by digitaldan** [comment]
> @digitaldan: Do you have some ideas?

err, i don't use this binding and don't think i have contributed to it so i'm not sure how to help.  

( i sometimes mix up threads about this binding and the openHAB Cloud binding since they sound similar in name, maybe thats the case here?)

**Comment by TDI4** [comment]
My solution was to install Home Assistant in my docker, installed iCloud integration, and established communication between the two systems via MQTT, and automations in HA. For example if me, as a person changing location, HA sends ON/OFF to my atHome, atWork switches. Or when my battery level changes, HA sends the value, and OH rule can warn me to charge my phone/watch.

It was urgent to solve the issue, because lot of my rules depending on if somebody is at home, or not. You can do that in less than 2 hours.

**Comment by JensHoRi** [comment]
OwnTracks with mqtt works fine too.

**Comment by BigGeorgeTx** [comment]
I am using OwnTracks with the GPSTracker Binding, which works well, perhaps better with my and my wife’s iPhone.  Better because you can set a geofence and get real-time presence versus the potential 5 minute or more delay with iCloud. 

But I use iCloud for devices like my AppleWatch that do not, at least as far as I know, support OwnTracks.

**Comment by mikthemonster** [comment]
Does anyone know where to find information about the "official" way to connect to iCloud, if we should re-write the binding?

**Comment by peteraquino** [comment]
Might be able to use the picklepete's pyicloud source code as a guide.

**Comment by mikthemonster** [comment]
I looked at picklepete's pyicloud and got it running on my Windows machine in Visual Studio Code. I can trigger "Find my iPhone" and get location etc. It is actually not very complicated.

I think it is "easy" to migrate the authentication principles to the the existing OpenHAB iCloud binding, but I dont have any clue about how to make or test an OpenHAB binding. I can probably make a Postman Collection showing all the principles and endpoints, but if some one could make the changes in the binding-code, then I would like to help making the changes - we could work together over Teams or something.

**Comment by mikthemonster** [comment]
We could also pay someone from freelancer to do it....

**Comment by maihacke** [comment]
If you could post the required HTTP request (inlcuding headers) here, I could have a look and try to create a PR + patched jar. The current implementation doesn't look too complicated.

**Comment by mikthemonster** [comment]
Hi Maihacke.

First Endpoint looks something like this (replace all the xxx with your own values):

```
HttpResponse<String> response = Unirest.post("https://idmsa.apple.com/appleauth/auth/signin?isRememberMeEnabled=true")
  .header("X-Apple-OAuth-Client-Id", "d39ba9916b7251055b22c7f910e2ea796ee65e98b2ddecea8f5dde8d9d1a815d")
  .header("X-Apple-OAuth-Client-Type", "firstPartyAuth")
  .header("X-Apple-OAuth-Redirect-URI", "https://www.icloud.com")
  .header("X-Apple-OAuth-Require-Grant-Code", "true")
  .header("X-Apple-OAuth-Response-Mode", "web_message")
  .header("X-Apple-OAuth-Response-Type", "code")
  .header("X-Apple-OAuth-State", "auth-xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx") //Generate code - I think it can anything
  .header("X-Apple-Widget-Key", "d39ba9916b7251055b22c7f910e2ea796ee65e98b2ddecea8f5dde8d9d1a815d")
  .header("Content-Type", "application/json")
  .body("{\r\n\"accountName\":\"xxx@xxx.xxx\",\r\n\"password\":\"xxxxxxxxxx\",\r\n\"rememberMe\":\"True\",\r\n\"trustTokens\": []\r\n}")
  .asString();

```

**Comment by maihacke** [comment]
Ok, what about the clientId. According to the docs it is something we need to acquire from WWDC:

https://developer.apple.com/documentation/sign_in_with_apple/clientconfigi/3230948-clientid

Is it okay to use the one from pyicloud?

**Comment by mikthemonster** [comment]
Hi Maihacke.

The ClientId from pyicloud is created using a function "uuid1" in python. It says: "uuid1(), uuid3(), uuid4(), uuid5() for generating version 1, 3, 4, and 5 UUIDs as specified in RFC 4122." 

Unfortunately Apple has blocked sending more authenticationcodes to my iPhone at the moment - i cant test if other uuid's are working...

**Comment by mikthemonster** [comment]
Maihacke -> If you like we could make a teams meeting and I can show you the pyicloud in action. Then I guess it is easier for you to understand what it takes to implement in OH.

**Comment by maihacke** [comment]
The clientId is hardcoded here: https://github.com/picklepete/pyicloud/blob/332cc9fa767862480c27253233c2cfdf9f2ea0d9/pyicloud/base.py#L392

The state is generated, which is compliant with the docs. Regarding meeting, I'm currently very busy... Hope I can contribute here something nevertheless.

**Comment by mikthemonster** [comment]
Ok, there are two different client id's I was talking about client id in line 398 from your link above.

**Comment by maihacke** [comment]
I was able to send the mentioned request from within the binding. But this triggers a popup for 2FA on my local mac. I had a look on the pyicloud code and my impression is, that there is a lot more to do to handle 2FA correctly...

**Comment by mikthemonster** [comment]
Yes, im working on the next requests - It takes some time because Apple will only send a few codes to my phone pr. day, and then I have to wait.

**Comment by mikthemonster** [comment]
Again, if we could look at it together it would be easier I think.

**Comment by maihacke** [comment]
From the community forum I found that somebody already create a java port from pyicloud. That would be possible of great help, if it works ;-): https://github.com/tmyroadctfig/icloud4j
I'm not sure if it supports 2FA and how the input of the code could be integrated in the openhab flow, since there is no way of user interaction when starting the binding.

**Comment by maihacke** [comment]
I started to create a proof-of-concept for including icloud4j into openhab-addons here https://github.com/maihacke/openhab-addons/tree/icloud

**Comment by maihacke** [comment]
I'm still struggling with the 2FA, I fear the library does not support it well. You have to to the login flow in one step. Login via username and password and then do the 2FA on the same instance. The only example I found tells to pause the programm flow via debugger. For sure that is no real solution. I cannot find a solution were you can start the authentication flow and then later on continue it when the user had entered the provided key. If this cannot be solved, we have to re-implement the icloud api interaction, based on pyicloud, the work from @mikthemonster and icloud4j.

**Comment by maihacke** [comment]
> https://www.getpostman.com/collections/407d255ecc18cdc8261c
> 
> 1.

I tried to use the collection to get some insides on the API. Login seems to work. I receive 204 after step 6 (enter token). But 7 is not working (450). Any ideas?

**Comment by mikthemonster** [comment]
Hi,

"204" after step 6 (enter token), is as expected. I get the same and it is working in my end.

Regarding "FindMyIphone_1_RefreshClient": This should return Status 200. Did you check the first part of the URL? Mine is: "https://p57-fmipweb.icloud.com:443" taken from "Auth_2_ListServices":

 "findme": {
            "url": "https://p57-fmipweb.icloud.com:443",
            "status": "active"
        },

And is it "status": "active"?

**Comment by maihacke** [comment]
The status was "active", the URL was different and I replaced it...

**Comment by maihacke** [comment]
Retried the postman collection  today. Now it is working. :-D

**Comment by mikthemonster** [comment]
That's great!

I was thinking if I could help making the changes to the  binding. What does it take (software etc.)? I have a Windows machine and is used to Visual Studio and VS Code. Is it easy to get started?

**Comment by maihacke** [comment]
If you know java it is not too complicated. I pushed my current experiments here:

https://github.com/maihacke/openhab-addons/tree/icloud-2

I dropped the use of icloud4j. I think it is not suitable without major changes.
Currently I try to reimplement authentication based on the postman collection, pyicloud and other sources.
The main class is ICloudConnectionV2.
You can run it from TestICloud testcase
You need to provide ICloud credentials via system properties.

This is currently just a proof of concept to see if I can get beyond 2FA authentication. Currently it is not working. Playing a sound is blocked (http 450).

If you like to play around run the test TestICloud#testAuth.

I will try to find out, why auth is working in postman but not in java.

**Comment by mikthemonster** [comment]
> I will try to find out, why auth is working in postman but not in java.

Postman includes all the cookies automatically - Im not sure how it is handled by the Java code.

**Comment by bananajoe86** [comment]
It sounds like there is a working solution. Maybe someone would like to write the walkthrough for dummies together. I'm not a newbie but I've never heard of postman and I have no idea what to do with it on my raspberry system.
thank you in advance for your support.
joe

**Comment by mikthemonster** [comment]
> It sounds like there is a working solution
Unfortunately there is no working solution at the moment. We have investigated how to log in to iCloud etc. at got that part working in another program (postman). Now someone has to make a binding for OpenHab.

**Comment by maihacke** [comment]
Just a minor update, I tried to create a PoC by reimplementing the necessary parts of pyicloud in java. My current results are here: https://github.com/maihacke/openhab-addons/tree/icloud-2
The authentication is working (in theory), so it accepts a verifycation code for 2FA. Im struggling to establish the trust to allow further request to the API. Which is more or less the same problem I had with postman. I will analyze this, if I find the time.

**Comment by maihacke** [comment]
If somebody has time he could try to help and find out, why the last step (getDevice) is failing.
What you have to do:
1. Checkout my branch icloud-2 (see above)
2. Run TestICloud#testAuth. I do this from eclipse. You have to provide your credentials in system variables, e.g. -Djdk.httpclient.HttpClient.log=all -Dicloud.test.email=XXXX -Dicloud.test.pw=XXXX
3. You will see a lot of debugging output. Many things are working. But the last step fetching the device infos (https://p61-fmipweb.icloud.com:443/fmipservice/client/web/refreshClien) fails with HTTP 450. I tried to compare to the requests from pyicloud (which are working for me). But up to now I couldn't find a reason.

**Comment by peteraquino** [comment]
I'm getting the same error (450), which also occurs in the Postman collection. Were you able to get around it in Postman?

**Comment by maihacke** [comment]
Yes, when I call 2_listdevices (which calls accountLogin) before the refresh devices.Am 04.11.2022 um 15:43 schrieb Peter Aquino ***@***.***>:﻿
I'm getting the same error (450), which also occurs in the Postman collection. Were you able to get around it in Postman?

**Comment by maihacke** [comment]
I created a minimal java snipped which mimics the behaviour:

```
    CookieManager cookieManager = new CookieManager();
    HttpClient client = HttpClient.newBuilder().version(Version.HTTP_1_1).followRedirects(Redirect.NORMAL)
        .connectTimeout(Duration.ofSeconds(20)).cookieHandler(cookieManager).build();
    String request = "{\"dsWebAuthToken\":\"--insert token--",\"extended_login\":true,\"accountCountryCode\":\"DEU\",\"trustToken\":\"\"}";
    client.send(
        HttpRequest.newBuilder().POST(BodyPublishers.ofString(request))
            .uri(URI.create("https://setup.icloud.com/setup/ws/1/accountLogin"))
            .headers("Origin", "https://www.icloud.com", "Referer", "https://www.icloud.com").build(),
        BodyHandlers.ofString());

    request = "{\"clientContext\":{\"deviceListVersion\":1,\"shouldLocate\":true,\"fmly\":false,\"selectedDevice\":\"all\"}}";
    client.send(
        HttpRequest.newBuilder().POST(BodyPublishers.ofString(request))
            .uri(URI.create("https://p61-fmipweb.icloud.com:443/fmipservice/client/web/refreshClient"))
            .headers("Origin", "https://www.icloud.com", "Referer", "https://www.icloud.com").build(),
        BodyHandlers.ofString());
```

**Comment by peteraquino** [comment]
Here's something that I'm remembering about my previous iCloud integration, which used pyicloud.  When the script launched, I got the 2FA notification (with code) on my trusted device, but I wasn't actually required to enter/submit the challenge code. The script would just continue to run and gather my device coordinates as expected.

Now, what I'm seeing with the Postman collection is that if I call Auth_1_UN/PW, I can select either "Don't Allow" or "Allow" on my trusted device...it doesn't really matter.  Then if I immediately call Auth_2_ListService, I get a good (200) response back containing all of my devices.  I can then call the refreshClient endpoint which returns a 200 and the associated details.

So it seems that by calling these endpoints, in this order, it isn't enforcing the 2FA at all.  A correct username and password is all that is required to retrieve device details (including location).

**Comment by maihacke** [comment]
Yes, that could be true. The Auth_1_UN/PW request should already set the X-Apple-ID-Session-Id (See Tests) which is used in Auth_2_ListServices to login. So Auth_3_EnterToken might not be necessary. But I for me the problems still remains that *_RefreshClient gives 450.

**Comment by peteraquino** [comment]
Let me take a look at that now and see if I get the same result.

**Comment by maihacke** [comment]
Good news. I was able to identify and solve the problem. The root cause is javas implementation of CookieManager, which leaves out the quotes (") when building the cookie header. I was able to create a solution for that. I have to clean it up a bit (remove private cookies for testing). Then I will push it and we can go forward.

**Comment by maihacke** [comment]
I made some more progress. I was able to integrate the new solution into the existing binding. Every thing is still very unpolished, but it works. If you like to try out feel free to do so. The commit id is db3b6b930f7e39d6d3711cbdd4a26b9133231bdf.
If you start with the new version you should receive a 2-FA code on one of your devices. Put this code into your thing config:

Bridge icloud:account:simon [appleId="*@*.*", password="***", code="XXXXXX", refreshTimeInMinutes=5]

DISCLAIMER: This solution is far from finished. There is no error handling. Sometimes codes are requested multiple times, which is confusing, and many more, ... but it is start...

**Comment by Flole998** [comment]
@maihacke If you want your fix merged at some point you should make sure you only catch the exception you want to catch and not just *ALL* exceptions.

**Comment by maihacke** [comment]
> @maihacke If you want your fix merged at some point you should make sure you only catch the exception you want to catch and not just _ALL_ exceptions.

Yeah sure, there are many issues like this in the current implementation. My approach was to get something running and clean it up afterwards.

**Comment by maihacke** [comment]
I created a new version here https://github.com/maihacke/openhab-addons/releases/tag/rc-1, I plan to integrate this into the official release in the future, so I would be glad if people could test it.

**Comment by BigGeorgeTx** [comment]
@Matsuo3rd Are you sure you are need 4 separate location refreshes?  I believe if you request a refresh, you get the latest location iCloud has for each device.  Sending a refresh request does not cause Apple to refresh the actual device location, it just sends you the latest location it has, or at least that it what I found when I studied the Last Location Update time stamps for my family's devices.  

Also, what are you using for the geofence trigger?  I'm currently using OwnTracks.

**Comment by Matsuo3rd** [comment]
@BigGeorgeTx: My experience shows that I actually need to run it twice for each device, waiting 20 seconds in between, in order to fetch "fresh/current" location data. This has proven to be deliver consistent results over time (2 years) - using iCloud binding.

Geofence is triggered through iOS geofence native feature leveraging homebridge + openHab virtual switches.
Virtual/dummy "geofence departure" and "geofence arrival" switches are exposed in homebridge / iOS HomeKit. A HomeKit geofence automation triggers those switches accordingly upon geofence triggers.
See https://community.openhab.org/t/presence-detection-close-to-home/84494/3

**Comment by Matsuo3rd** [comment]
Some feedback on RC-1

At openHab startup, a harmless message shows up. Still the binding works.
However, warning logs always make me nervous (especially at startup time).

[2022-11-15 15:12:28] [WARN ] Need to authenticate first.
java.lang.IllegalStateException: Webservice URLs not set. Need to authenticate first.
	at org.openhab.binding.icloud.internal.ICloudService.getDevices(ICloudService.java:293) ~[?:?]
	at org.openhab.binding.icloud.internal.handler.ICloudAccountBridgeHandler.lambda$2(ICloudAccountBridgeHandler.java:128) ~[?:?]
	at org.openhab.core.cache.ExpiringCache.refreshValue(ExpiringCache.java:101) ~[?:?]
	at org.openhab.core.cache.ExpiringCache.getValue(ExpiringCache.java:72) ~[?:?]
	at org.openhab.binding.icloud.internal.handler.ICloudAccountBridgeHandler.refreshData(ICloudAccountBridgeHandler.java:280) ~[?:?]
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:515) [?:?]
	at java.util.concurrent.FutureTask.runAndReset(FutureTask.java:305) [?:?]
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:305) [?:?]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128) [?:?]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628) [?:?]
	at java.lang.Thread.run(Thread.java:829) [?:?]

**Comment by Matsuo3rd** [comment]
Another feedback on RC-1

When launching simultaneous calls / or say 2/3 sequential calls without enough delay in between, iCloud service returns 450 error. I suspect this is an anti-spam measure. However, the binding does not recover from that, even after waiting for couple of hours. One need to restart openHab to get the binding back in business.

Maybe the binding needs to reconnect (or execute some logic) after such error.

[2022-11-15 15:02:17] [WARN ] Unable to refresh device data
org.openhab.binding.icloud.internal.ICloudAPIResponseException: Request https://p105-fmipweb.icloud.com:443/fmipservice/client/web/refreshClient failed with 450.
	at org.openhab.binding.icloud.internal.ICloudSession.request(ICloudSession.java:142) ~[?:?]
	at org.openhab.binding.icloud.internal.ICloudSession.post(ICloudSession.java:97) ~[?:?]
	at org.openhab.binding.icloud.internal.FindMyIPhoneServiceManager.refreshClient(FindMyIPhoneServiceManager.java:69) ~[?:?]
	at org.openhab.binding.icloud.internal.handler.ICloudAccountBridgeHandler.lambda$2(ICloudAccountBridgeHandler.java:128) ~[?:?]
	at org.openhab.core.cache.ExpiringCache.refreshValue(ExpiringCache.java:101) [bundleFile:?]
	at org.openhab.core.cache.ExpiringCache.getValue(ExpiringCache.java:72) [bundleFile:?]
	at org.openhab.binding.icloud.internal.handler.ICloudAccountBridgeHandler.refreshData(ICloudAccountBridgeHandler.java:280) [bundleFile:?]
	at org.openhab.binding.icloud.internal.handler.ICloudDeviceHandler.handleCommand(ICloudDeviceHandler.java:159) [bundleFile:?]
	at jdk.internal.reflect.GeneratedMethodAccessor144.invoke(Unknown Source) ~[?:?]
	at jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[?:?]
	at java.lang.reflect.Method.invoke(Method.java:566) ~[?:?]
	at org.openhab.core.internal.common.AbstractInvocationHandler.invokeDirect(AbstractInvocationHandler.java:154) [bundleFile:?]
	at org.openhab.core.internal.common.InvocationHandlerSync.invoke(InvocationHandlerSync.java:59) [bundleFile:?]
	at com.sun.proxy.$Proxy516.handleCommand(Unknown Source) [?:?]
	at org.openhab.core.thing.internal.profiles.ProfileCallbackImpl.handleCommand(ProfileCallbackImpl.java:80) [bundleFile:?]
	at org.openhab.core.thing.internal.profiles.SystemDefaultProfile.onCommandFromItem(SystemDefaultProfile.java:48) [bundleFile:?]
	at jdk.internal.reflect.GeneratedMethodAccessor143.invoke(Unknown Source) ~[?:?]
	at jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[?:?]
	at java.lang.reflect.Method.invoke(Method.java:566) ~[?:?]
	at org.openhab.core.internal.common.AbstractInvocationHandler.invokeDirect(AbstractInvocationHandler.java:154) [bundleFile:?]
	at org.openhab.core.internal.common.Invocation.call(Invocation.java:52) [bundleFile:?]
	at java.util.concurrent.FutureTask.run(FutureTask.java:264) [?:?]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128) [?:?]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628) [?:?]
	at java.lang.Thread.run(Thread.java:829) [?:?]

**Comment by maihacke** [comment]
> Some feedback on RC-1
> 
> At openHab startup, a harmless message shows up. Still the binding works. However, warning logs always make me nervous (especially at startup time).
> 
> ```
> [2022-11-15 15:12:28] [WARN ] Need to authenticate first.
> java.lang.IllegalStateException: Webservice URLs not set. Need to authenticate first.
> 	at org.openhab.binding.icloud.internal.ICloudService.getDevices(ICloudService.java:293) ~[?:?]
> 	at org.openhab.binding.icloud.internal.handler.ICloudAccountBridgeHandler.lambda$2(ICloudAccountBridgeHandler.java:128) ~[?:?]
> 	at org.openhab.core.cache.ExpiringCache.refreshValue(ExpiringCache.java:101) ~[?:?]
> 	at org.openhab.core.cache.ExpiringCache.getValue(ExpiringCache.java:72) ~[?:?]
> 	at org.openhab.binding.icloud.internal.handler.ICloudAccountBridgeHandler.refreshData(ICloudAccountBridgeHandler.java:280) ~[?:?]
> 	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:515) [?:?]
> 	at java.util.concurrent.FutureTask.runAndReset(FutureTask.java:305) [?:?]
> 	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:305) [?:?]
> 	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128) [?:?]
> 	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628) [?:?]
> 	at java.lang.Thread.run(Thread.java:829) [?:?]
> ```

Nothing to get nervous. Even if the explanation is in the message, I lowered the log level to debug. So it wont show up any more. I think there are enough hints in the log and UI that you need to login.

**Comment by maihacke** [comment]
> Another feedback on RC-1
> 
> When launching simultaneous calls / or say 2/3 sequential calls without enough delay in between, iCloud service returns 450 error. I suspect this is an anti-spam measure. However, the binding does not recover from that, even after waiting for couple of hours. One need to restart openHab to get the binding back in business.
> 
> Maybe the binding needs to reconnect (or execute some logic) after such error.
> 
> ```
> [2022-11-15 15:02:17] [WARN ] Unable to refresh device data
> org.openhab.binding.icloud.internal.ICloudAPIResponseException: Request https://p105-fmipweb.icloud.com:443/fmipservice/client/web/refreshClient failed with 450.
> 	at org.openhab.binding.icloud.internal.ICloudSession.request(ICloudSession.java:142) ~[?:?]
> 	at org.openhab.binding.icloud.internal.ICloudSession.post(ICloudSession.java:97) ~[?:?]
> 	at org.openhab.binding.icloud.internal.FindMyIPhoneServiceManager.refreshClient(FindMyIPhoneServiceManager.java:69) ~[?:?]
> 	at org.openhab.binding.icloud.internal.handler.ICloudAccountBridgeHandler.lambda$2(ICloudAccountBridgeHandler.java:128) ~[?:?]
> 	at org.openhab.core.cache.ExpiringCache.refreshValue(ExpiringCache.java:101) [bundleFile:?]
> 	at org.openhab.core.cache.ExpiringCache.getValue(ExpiringCache.java:72) [bundleFile:?]
> 	at org.openhab.binding.icloud.internal.handler.ICloudAccountBridgeHandler.refreshData(ICloudAccountBridgeHandler.java:280) [bundleFile:?]
> 	at org.openhab.binding.icloud.internal.handler.ICloudDeviceHandler.handleCommand(ICloudDeviceHandler.java:159) [bundleFile:?]
> 	at jdk.internal.reflect.GeneratedMethodAccessor144.invoke(Unknown Source) ~[?:?]
> 	at jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[?:?]
> 	at java.lang.reflect.Method.invoke(Method.java:566) ~[?:?]
> 	at org.openhab.core.internal.common.AbstractInvocationHandler.invokeDirect(AbstractInvocationHandler.java:154) [bundleFile:?]
> 	at org.openhab.core.internal.common.InvocationHandlerSync.invoke(InvocationHandlerSync.java:59) [bundleFile:?]
> 	at com.sun.proxy.$Proxy516.handleCommand(Unknown Source) [?:?]
> 	at org.openhab.core.thing.internal.profiles.ProfileCallbackImpl.handleCommand(ProfileCallbackImpl.java:80) [bundleFile:?]
> 	at org.openhab.core.thing.internal.profiles.SystemDefaultProfile.onCommandFromItem(SystemDefaultProfile.java:48) [bundleFile:?]
> 	at jdk.internal.reflect.GeneratedMethodAccessor143.invoke(Unknown Source) ~[?:?]
> 	at jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[?:?]
> 	at java.lang.reflect.Method.invoke(Method.java:566) ~[?:?]
> 	at org.openhab.core.internal.common.AbstractInvocationHandler.invokeDirect(AbstractInvocationHandler.java:154) [bundleFile:?]
> 	at org.openhab.core.internal.common.Invocation.call(Invocation.java:52) [bundleFile:?]
> 	at java.util.concurrent.FutureTask.run(FutureTask.java:264) [?:?]
> 	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128) [?:?]
> 	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628) [?:?]
> 	at java.lang.Thread.run(Thread.java:829) [?:?]
> ```

Did you have this messages repeated in the logs, or did the bridge just stay in OFFLINE?

**Comment by maihacke** [comment]
> The RC-1 looks promising. I managed to get it running. Thank you very much for taking the lead!
> 
> However, the plugin stopped working after a while (450 errors) along with multiple email sent from Apple notifying me that "Your Apple ID was used to sign into Icloud via a Web browser".
> 
> My experience shows that if the iCloud web service is called multiple times in parallel, Apple triggers this. My use case is to launch a location refresh for each of my family Apple devices (4 of them) upon geofence arrive/departure trigger from an openHab rules.
> 
> My previous openHab/pyicloud workaround was to lock / synchronize parallel calls in order to prevent this. i.e. wait for the previous one to complete before launching another call. Would it be something to be considered in the binding - or shall this "lock" mechanism be implemented in the openHab rules?
> 
> Btw, shall we report issue here? or somewhere else?
> 
> Again, thanks for you effort @maihacke

I currently have no plans to include a lock feature in the binding and try to follow KISS principle :-)

**Comment by Matsuo3rd** [comment]
> Was this messages repeated in the logs, or did the bridge just stay in OFFLINE?

Both. The exact same error message is repeated each time I execute a REFRESH call for location data.
And the iCloud Bridge / Account Thing is marked as OFFLINE (CONFIGURATION_ERROR Invalid credentials).
Configuration / credentials are ok, because just restarting openHab puts the Bridge back on green / ONLINE.

**Comment by Matsuo3rd** [comment]
> I currently have no plans to include a lock feature in the binding and try to follow KISS principle :-)

I totally get that. May be, depending on the code architecture, a `synchronized` keyword on the calling method might be enough.

**Comment by maihacke** [comment]
> > Was this messages repeated in the logs, or did the bridge just stay in OFFLINE?
> 
> Both. The exact same error message is repeated each time I execute a REFRESH call for location data. And the iCloud Bridge / Account Thing is marked as OFFLINE (CONFIGURATION_ERROR Invalid credentials). Configuration / credentials are ok, because just disabling/enabling Thing or restarting openHab puts the Bridge back on green / ONLINE.

Could you please check if pausing and reenabling the bridge thing has the same effect?

**Comment by Matsuo3rd** [comment]
> Could you please check if pausing and reenabling the bridge thing has the same effect?

Yes it has the same effect (i.e. makes the Thing back Green/ONLINE). That was actually what I meant by "disabling/enabling Thing".

**Comment by Matsuo3rd** [comment]
> @Matsuo3rd Are you sure you are need 4 separate location refreshes?

@BigGeorgeTx : you were right. One call is enough to grab every devices' location. So I don't need to execute one call for each device. However, I still keep executing a second call 20 seconds later to collect current data.

**Comment by maihacke** [comment]
> > Could you please check if pausing and reenabling the bridge thing has the same effect?
> 
> Yes it has the same effect (i.e. makes the Thing back Green/ONLINE). That was actually what I meant by "disabling/enabling Thing".

Ah sorry, overlooked that you already tried. Can you post a debug log here, so that I can see what happens?

**Comment by Matsuo3rd** [comment]
See iCloud.log file attached w/ DEBUG level set on.
Btw, I re-enable programmatically the iCloud Bridge into a rule before executing REFRESH action (if status is OFFLINE)

[icloud.log](https://github.com/openhab/openhab-addons/files/10040236/icloud.log)

**Comment by maihacke** [comment]
Thanks for the log. The problem is that currently the login is decoupled from the refresh. This is due to the complexity of asking the user for the 2fa key and avoid addtional auth request to allow the user to actually enter the key. I'm pretty sure the problem had been resolved by itself after one hour, when the authentication is refreshed. I'm aware that is not a optimal solution and leaves room for future improvement.

**Comment by Matsuo3rd** [comment]
OK. I am good with that given that I came out with a workaround.

**Comment by BigGeorgeTx** [comment]
> > @Matsuo3rd Are you sure you are need 4 separate location refreshes?
> 
> @BigGeorgeTx : you were right. One call is enough to grab every devices' location. So I don't need to execute one call for each device. However, I still keep executing a second call 20 seconds later to collect current data.

@Matsuo3rd Glad to hear that worked for you.  The second call after 20 seconds works great!  Thanks for the tip.

**Comment by RalphSester** [comment]
Hello everybody,

with new modified binding of @maihacke , everything is working fine for me. Thank you very much for this work.

But every day the binding is going offline and has to be restarted at the moment.
I have written a script, which automates this.....

Everytime when i restart the binding i receive an email from Apple, that "Your Apple-ID is used to login with a browser".
Are their plans (or is it possible) to change something to renew the Auth-Token without these struggles ?

Thank You!
Ralph

**Comment by maihacke** [comment]
> Hello everybody,
> 
> with new modified binding of @maihacke , everything is working fine for me. Thank you very much for this work.
> 
> But every day the binding is going offline and has to be restarted at the moment. I have written a script, which automates this.....
> 
> Everytime when i restart the binding i receive an email from Apple, that "Your Apple-ID is used to login with a browser". Are their plans (or is it possible) to change something to renew the Auth-Token without these struggles ?
> 
> Thank You! Ralph

This is not the normal behaviour. Probably something with the 2-FA went wrong or you haven't 2-FA activated for your iCloud account.

**Comment by RalphSester** [comment]
> This is not the normal behaviour. Probably something with the 2-FA went wrong or you haven't 2-FA activated for your iCloud account.

I have 2-FA activated, but everytime, when i start the binding i get this email......
In my Things-File i have no "code" defined in the bridge-thing....

Can i check something to find the problem?

**Comment by maihacke** [comment]
That's strange normally you should receive a code on one of your devices, which you have to provide in the config to prevent further logins/mails. You should check the logs, what happens here. Alternativly you could try to "reset" the bridge by removing it and re-adding it to openhab.

**Comment by BigGeorgeTx** [comment]
I'm seeing the same things as @RalphSester. Every day or so, the iCloud binding goes Offline with the 450 error.  If a cycle the iCloud Account Thing in the UI, it comes right back on line.  And then I get an email telling me that my Apple ID was used to sign into iCloud.

Ralph:  can you share your code for automatically restarting the iCloud account thing when it goes offline?  Thanks

**Comment by RalphSester** [comment]
> Ralph: can you share your code for automatically restarting the iCloud account thing when it goes offline? Thanks

ok @BigGeorgeTx , here is my solution:

I use the following rule to detect, if the Bridge-Thing comes offline:

```
rule "Bridge_Apple_Update"
when
   Thing "icloud:account:XXXXX" changed
then
   var thingStatusInfo = getThingStatusInfo("icloud:account:XXXXX")
   var String status = thingStatusInfo.getStatus().toString()
   switch(status) {
      case "ONLINE": {
               //Do some stuff if you want
                      }
      case "OFFLINE": {
               executeCommandLine ("Putty/OpenPutty.bat", "RestartApple.txt")
                      }
   }
end

```
The command line starts a batch-file, which is starting putty.exe with the "RestartApple.txt" as script.
The script contains the following command:

`bundle:restart 'openHAB Add-ons :: Bundles :: iCloud Binding'`

All the Files (putty.exe, the batch-file and the RestartApple.txt) are store in a directory "Putty" in the "userdata"-Directory of OpenHAB. I am using OpenHAB in a Windows Server Environment. May be, that there are some differences to Raspi or other plattforms. 

Ralph

**Comment by JensHoRi** [comment]
Restarting the full binding not necessary is my experience. Wait (normally bridge comes offline in a few minutes again) or disable and enable the bridge thing. This login again without a new 2FA, but creates a mail about new login via apple

**Comment by RalphSester** [comment]
> Restarting the full binding not necessary

Yes, i know. But to disable and re-enable the bridge-thing is not automateable....therefore i restart the binding (automatically)

**Comment by JensHoRi** [comment]
That’s not true. You can do via rest-API (enable/disable)

**Comment by maihacke** [comment]
As already said the current auth flow is not optimal. I suggest to not put to much effort into automating workarounds, since the behaviour will be changed in the future.

**Comment by maihacke** [comment]
I created a new RC https://github.com/maihacke/openhab-addons/releases/tag/rc-2
Please test and provide feedback.

**Comment by RalphSester** [comment]
I have installed RC2 yesterday....with no problems until now.
Now I have installed RC3 and will report, if I still receive the notification emails from apple every day....
Thank you for your work!

**Comment by RalphSester** [comment]
Hi @maihacke ,

after more than 24 hours since I installed your RC3, i can report, that the binding does not going offline like before and it seems that the renewal of the icloud-session-token is working transparent for users.

It seems, that your work is exellent!

Ralph

**Comment by maihacke** [comment]
Thanks, that are good news. Hopefully this could be integrated into the next OH release which is very close.

**Comment by maihacke** [comment]
A lot of work as been done to prepare merge into main line. Code should work as before, but one can never know. So please test the new https://github.com/maihacke/openhab-addons/releases/tag/rc-4.

**Comment by FlorianReinartz** [comment]
Hello everyone,

I installed and set up the RC4.
Worked right away.
I will report if there are problems.

Thank you so much for doing the binding programming.

The configuration of the bindings also works wonderfully as a text file.
This is my configuration "icloud.things" for those who don't do the configuration via UI.
`
Bridge icloud:account:iCloud_FR "iCloud Bridge"             [appleId="XXXXXX@XXX.XXX password="XYZ123", refreshTimeInMinutes=5, code="########"]
{
    Thing device iPhone_Flori   "iCloud Florians iPhone"    [deviceId="JGFKUZGLZGB"]
    Thing device iPhone_Tine    "iCloud Tines iPhone"       [deviceId="JGFKUZGLZGB"]
    Thing device iPhone_Johanna "iCloud Johannas iPhone"    [deviceId="JGFKUZGLZGB"]
}
`

Regards
Florian

**Comment by maihacke** [comment]
Here is another one for testing: https://github.com/maihacke/openhab-addons/releases/tag/rc-5

**Comment by RalphSester** [comment]
jo....installed now....comes online....retrieving all data.....will report after 24 hours again!

**Comment by maihacke** [comment]
Here is another one for testing: https://github.com/maihacke/openhab-addons/releases/tag/rc-6

(you do not have to follow all RCs for sure, but if people join, the should always use the latest one)

**Comment by maihacke** [comment]
The PR was approved. This is a RC for the version, which is going to be merged. If you find the time please test it:

https://github.com/maihacke/openhab-addons/releases/tag/rc-7

**Comment by RalphSester** [comment]
Thanks everybody !!!

**Comment by FlorianReinartz** [comment]
Hello,
I've been testing the RC4 version for 40 hours now.
Runs without problems so far.
Connects again after a reboot without any problems.

**Comment by Jan1503** [comment]
Hey!
May I kindly ask for the postman-collection of the sign-in process?
The old link is expired :(
If I ever find some free time, I would like to port the stuff to .net.

Thanks!

**Comment by maihacke** [comment]
Sorry switches my notebook inbetween. Doesn't have it either. You could also think of using the java code as a blueprint, starting from org.openhab.binding.icloud.TestICloud#testAuth. It's not to complicated.
